mindsdb_native/libs/helpers/confidence_helpers.py

```diff
@@ -66,39 +66,49 @@ def set_conf_range(X, icp, target, typing_info, lmd, std_tol=1, group='__default
-    significances = []
-    conf_ranges = []
-    std_dev = stats[predicted_col]['train_std_dev'][group]
-
-    tolerance = std_dev * std_tol
-
-    for sample_idx in range(all_confs.shape[0]):
-        sample = all_confs[sample_idx, :, :]
-        for idx in range(sample.shape[1]):
-            significance = (99 - idx) / 100
-            diff = sample[1, idx] - sample[0, idx]
-            if diff <= tolerance:
-                conf_range = list(sample[:, idx])
-                if stats[predicted_col].get('positive_domain', False):
-                    conf_range[0] = max(0.0, conf_range[0])
-
-                significances.append(significance)
+def get_numerical_conf_range(all_confs, predicted_col, stats, std_tol=1, group='__default', error_rate=None):
+    """ Gets prediction bounds for numerical targets, based on ICP estimation and width tolerance
+        error_rate: pre-determined error rate for the ICP, used in anomaly detection tasks to adjust the
+        threshold sensitivity
+    """
+    if not isinstance(error_rate, float):
+        error_rate = None
+
+    if error_rate is None:
+        significances = []
+        conf_ranges = []
+        std_dev = stats[predicted_col]['train_std_dev'][group]
+
+        tolerance = std_dev * std_tol
+
+        for sample_idx in range(all_confs.shape[0]):
+            sample = all_confs[sample_idx, :, :]
+            for idx in range(sample.shape[1]):
+                significance = (99 - idx) / 100
+                diff = sample[1, idx] - sample[0, idx]
+                if diff <= tolerance:
+                    conf_range = list(sample[:, idx])
+                    significances.append(significance)
+                    conf_ranges.append(conf_range)
+                    break
+            else:
+                significances.append(0.9991)  # default: confident that value falls inside big bounds
+                bounds = sample[:, 0]
+                sigma = (bounds[1] - bounds[0]) / 4
+                conf_range = [bounds[0] - sigma, bounds[1] + sigma]
                 conf_ranges.append(conf_range)
-                break
-        else:
-            significances.append(0.9991)  # default: confident that value falls inside big bounds
-            bounds = sample[:, 0]
-            sigma = (bounds[1] - bounds[0]) / 4
-            conf_range = [bounds[0] - sigma, bounds[1] + sigma]
 
-            if stats[predicted_col].get('positive_domain', False):
-                conf_range[0] = max(0.0, conf_range[0])
+        conf_ranges = np.array(conf_ranges)
+    else:
+        # fixed error rate
+        conf = 1 - error_rate
+        conf_idx = int(100*error_rate) - 1
+        conf_ranges = all_confs[:, :, conf_idx]
+        significances = [conf for _ in range(conf_ranges.shape[0])]
 
-            conf_ranges.append(conf_range)
-
-    return significances, np.array(conf_ranges)
+    if stats[predicted_col].get('positive_domain', False):
+        conf_ranges[conf_ranges < 0] = 0
```

Comment:
I see youve extended the confidence ranges - since it looks like you're looking for all instances where conf_ranges < 0 to convert to 0

Reply:
or rather, how large can conf_ranges be?

Reply:
Depending on the nonconformity scores they can be quite big. Also, unless normalized they are symmetrical, which is why for a large enough bound you can get a partially negative prediction range for a target you know cannot be negative. That's why this filter is in place.